File: app/src/main/java/org/teamblueridge/pasteitapp/MainActivity.java
package org.teamblueridge.pasteitapp;

import android.app.FragmentManager;
import android.app.ProgressDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;

import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.message.BasicNameValuePair;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;


public class MainActivity extends ActionBarActivity {

    private static final String TAG = "TeamBlueRidge";

    String mPasteUrlString;
    String mUploadUrl;
    String mUploadingText;
    String mUserName;
    String mToastText;
    String mFileContents;
    Intent mReceivedIntent;
    String mReceivedAction;
    ProgressDialog pDialogFileLoad;
    private ProgressDialog pDialogUpload;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //We're using the toolbar from AppCompat as our actionbar
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
        }
        getSupportActionBar().getThemedContext();
        // Set-up the paste fragment and give it a name so we can track it
        if (savedInstanceState == null) {
            getFragmentManager().beginTransaction()
                    .add(R.id.container, new PasteFragment(), "PasteFragment")
                    .commit();
        }

        // Set-up up navigation
        getFragmentManager().addOnBackStackChangedListener(new FragmentManager
                .OnBackStackChangedListener() {
            @Override
            public void onBackStackChanged() {
                int stackHeight = getFragmentManager().getBackStackEntryCount();
                /* Check if we have anything on the stack. If we do, we need to have the back button
                 * display in in the ActionBar */
                if (stackHeight > 0) {
                    getSupportActionBar().setHomeButtonEnabled(true);
                    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
                } else {
                    getSupportActionBar().setDisplayHomeAsUpEnabled(false);
                    getSupportActionBar().setHomeButtonEnabled(false);
                }
            }

        });

        // Get tracker.
        Tracker t = ((Analytics) this.getApplication()).getTracker(
                Analytics.TrackerName.APP_TRACKER);

        // Set up the status bar tint using the carbonrom SysBarTintManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN
                & Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
            SysBarTintManager.setupTranslucency(this, true, false);

            SysBarTintManager mTintManager = new SysBarTintManager(this);
            mTintManager.setStatusBarTintEnabled(true);
            mTintManager.setActionBarTintEnabled(true);
            mTintManager.setStatusBarTintColor(getResources().getColor(R.color.blue_700));
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
            Window window = this.getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.setStatusBarColor(this.getResources().getColor(R.color.blue_700));
        }
    }

    public void onStart() {
        super.onStart();
        mReceivedIntent = getIntent();
        mReceivedAction = mReceivedIntent.getAction();
        EditText pasteContentEditText = (EditText) findViewById(R.id.paste_content_edittext);
        if (mReceivedAction.equals(Intent.ACTION_VIEW) || mReceivedAction.equals(Intent.ACTION_EDIT)) {
            String receivingText = getResources().getString(R.string.file_load);
            //Prepare the dialog for loading a file
            pDialogFileLoad = new ProgressDialog(MainActivity.this);
            pDialogFileLoad.setMessage(receivingText);
            pDialogFileLoad.setIndeterminate(true);
            pDialogFileLoad.setCancelable(false);
            pDialogFileLoad.show();
            new LoadFile().execute();
            if (pDialogFileLoad.isShowing()) {
                pDialogFileLoad.dismiss();
            }
            mReceivedAction = Intent.ACTION_DEFAULT;
        } else {
            String receivedText = mReceivedIntent.getStringExtra(Intent.EXTRA_TEXT);
            if (receivedText != null) {
                pasteContentEditText.setText(receivedText);
            }
        }

        // Check if the "languages" file exists
        File file = getBaseContext().getFileStreamPath("languages");
        if (file.exists()) {
            new PopulateSpinner().execute();
        } else {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(MainActivity.this);
            ApiHandler apiHandler = new ApiHandler();
            apiHandler.getLanguagesAvailable(prefs, this);
            new PopulateSpinner().execute();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        switch (id) {
            case R.id.action_paste:
                doPaste();
                return true;
            case R.id.action_settings:
                Intent intent = new Intent(this, SettingsActivity.class);
                startActivity(intent);
                return true;
            case android.R.id.home:
                getFragmentManager().popBackStack();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    /**
     * Does some preparation before finally calling the UploadPaste ASyncTask
     */
    public void doPaste() {
        EditText pasteNameEditText = (EditText) findViewById(R.id.paste_name_edittext);
        EditText pasteContentEditText = (EditText) findViewById(R.id.paste_content_edittext);
        String pasteContentString = pasteContentEditText.getText().toString();
        Spinner languageSpinner = (Spinner) findViewById(R.id.language_spinner);
        Integer languageSelected = languageSpinner.getSelectedItemPosition();
        ApiHandler apiHandler = new ApiHandler();
        String[] languageListStringArray =
                apiHandler.getLanguageArray(getApplicationContext(), "pretty");

        String language = languageListStringArray[languageSelected];
        if (!pasteContentString.isEmpty()) {
            if (NetworkUtil.isConnectedToNetwork(this)) {
                if (!language.equals("0")) {
                    new UploadPaste().execute(language);
                    mToastText = getResources().getString(R.string.paste_toast);
                    pasteNameEditText.setText("");
                    pasteContentEditText.setText("");
                } else {
                    mToastText = getResources().getString(R.string.invalid_language);
                }
            } else {
                mToastText = getResources().getString(R.string.no_network);
            }
        } else {
            mToastText = getResources().getString(R.string.paste_no_text);
        }
        Context context = getApplicationContext();
        CharSequence text = mToastText;
        int duration = Toast.LENGTH_SHORT;
        Toast.makeText(context, text, duration).show();
    }

    /**
     * Sets the title of the ActionBar to be whatever is specified. Called from the various
     * fragments and other activities
     *
     * @param title The title to be used for the ActionBar
     */
    public void setActionBarTitle(String title) {
        //Set the title of the action bar
        //Called from the fragments
        getSupportActionBar().setTitle(title);
    }

    /**
     * AsyncTask for loading the file, from an intent, into the paste content edittext</p>
     * <p/>
     * doInBackground : get the file's contents loaded</p>
     * onPostExecute : update the edittext's content</p>
     */
    class LoadFile extends AsyncTask<String, String, String> {

        protected String doInBackground(String... args) {
            try {
                mReceivedAction = getIntent().getAction();
                Uri receivedText = mReceivedIntent.getData();
                InputStream inputStream = getContentResolver().openInputStream(receivedText);
                BufferedReader reader = new BufferedReader(new InputStreamReader(
                        inputStream));
                StringBuilder stringBuilder = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    stringBuilder.append(line);
                }
                inputStream.close();
                mFileContents = stringBuilder.toString();
            } catch (IOException e) {
                Log.e(TAG, e.toString());

            }
            return mFileContents;
        }

        protected void onPostExecute(String file) {
            mFileContents = file;
            runOnUiThread(new Runnable() {
                public void run() {
                    //Create a clickable link from pasteUrlString for user (opens in web browser)
                    EditText pasteContentEditText;
                    pasteContentEditText = (EditText) findViewById(R.id.paste_content_edittext);
                    pasteContentEditText.setText(mFileContents);
                }
            });
        }
    }

    /**
     * ASyncTask that uploads the paste to the selected server</p>
     * <p/>
     * onPreExecute : Setup the upload dialog</p>
     * doInBackground : Perform the upload operation with an HttpClient</p>
     * onPostExecute : Close the dialog box and update the paste url label</p>
     */
    class UploadPaste extends AsyncTask<String, String, String> {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(MainActivity.this);
        TextView pasteUrlLabel = (TextView) findViewById(R.id.paste_url_label);
        EditText pasteNameEditText = (EditText) findViewById(R.id.paste_name_edittext);
        String pasteNameString = pasteNameEditText.getText().toString();
        EditText pasteContentEditText = (EditText) findViewById(R.id.paste_content_edittext);
        String pasteContentString = pasteContentEditText.getText().toString();
        final String HTTP_USER_AGENT = "Paste It v" + getString(R.string.version_name) +
                ", an Android app for pasting to Stikked " +
                "(https://play.google.com/store/apps/details?id=org.teamblueridge.pasteitapp)";

        @Override
        protected void onPreExecute() {
            mUploadingText = getResources().getString(R.string.paste_upload);
            super.onPreExecute();
            //Prepare the dialog for uploading the paste
            pDialogUpload = new ProgressDialog(MainActivity.this);
            pDialogUpload.setMessage(mUploadingText);
            pDialogUpload.setIndeterminate(false);
            pDialogUpload.setCancelable(false);
            pDialogUpload.show();
        }

        // post the content in the background while showing the dialog
        protected String doInBackground(String... args) {
            try {
                HttpClient httpclient = new DefaultHttpClient();
                UploadDownloadUrlPrep upDownPrep = new UploadDownloadUrlPrep();
                mUploadUrl = upDownPrep.prepUrl(prefs, "upCreate");
                String language = args[0];

                //Ensure username is set, if not, default to "mobile user"
                if (!prefs.getString("pref_name", "").isEmpty()) {
                    mUserName = prefs.getString("pref_name", "");
                } else {
                    mUserName = "Mobile User";
                }
                if (language == null || language.isEmpty()){
                    language = "plaintext";
                }
                //Get ready to actually send everything to the server
                HttpPost httppost = new HttpPost(mUploadUrl);
                // HTTP Header data
                httppost.setHeader("User-Agent", HTTP_USER_AGENT);
                List<NameValuePair> nameValuePairs = new ArrayList<>(2);
                nameValuePairs.add(new BasicNameValuePair("title", pasteNameString));
                nameValuePairs.add(new BasicNameValuePair("text", pasteContentString));
                nameValuePairs.add(new BasicNameValuePair("name", mUserName));
                nameValuePairs.add(new BasicNameValuePair("lang", language));
                httppost.setEntity(new UrlEncodedFormEntity(nameValuePairs));
                // Execute HTTP Post Request
                HttpResponse response = httpclient.execute(httppost);
                InputStream in = response.getEntity().getContent();
                StringBuilder stringbuilder = new StringBuilder();
                BufferedReader bfrd = new BufferedReader(new InputStreamReader(in), 1024);
                String line;
                while ((line = bfrd.readLine()) != null)
                    stringbuilder.append(line);
                mPasteUrlString = stringbuilder.toString();
            } catch (IOException e) {
                Log.e(TAG, e.toString());
            }
            return null;
        }

        //Since we used a dialog, we need to disable it
        protected void onPostExecute(String paste_url) {
            //Dismiss the dialog after getting all products
            pDialogUpload.dismiss();
            //Copy pasteUrl to clipboard
            ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("PasteIt", mPasteUrlString);
            clipboard.setPrimaryClip(clip);

            //Display paste URL if allowed in preferences
            runOnUiThread(new Runnable() {
                public void run() {
                    //Create a clickable link from pasteUrlString for user (opens in web browser)
                    String linkText = "<a href=\"" + mPasteUrlString + "\">" + mPasteUrlString + "</a>";
                    pasteUrlLabel.setText(Html.fromHtml(linkText));
                    pasteUrlLabel.setMovementMethod(LinkMovementMethod.getInstance());
                }
            });
        }
    }

    /**
     * Puts the languages from JSON grabbed via the Stikked API into the Spinner</p>
     * <p/>
     * doInBackground : Read from JSON file, turn JSON into an array of objects and an array of
     * keys.</p>
     * onPostExecute : Populate the spinner with the Array from JSON</p>
     */
    class PopulateSpinner extends AsyncTask<String, String, String[]> {
        ApiHandler apiHandler = new ApiHandler();

        @Override
        protected String[] doInBackground(String... params) {
            //Read the JSON file from internal storage
            return apiHandler.getLanguageArray(getApplicationContext(), "pretty");
        }

        @Override
        protected void onPostExecute(final String[] langListPretty) {
            //Populate spinner
            super.onPostExecute(langListPretty);
            runOnUiThread(new Runnable() {
                public void run() {
                    SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(MainActivity.this);
                    String positionListPref = prefs.getString("pref_default_language", "-1");
                    ArrayAdapter<String> uglyList = new ArrayAdapter<>
                            (getApplicationContext(), R.layout.spinner_item, apiHandler
                                    .getLanguageArray(getApplicationContext(), "ugly"));
                    try {
                        Spinner spinner = (Spinner) findViewById(R.id.language_spinner);
                        ArrayAdapter<String> adapter =
                                new ArrayAdapter<>(getApplicationContext(),
                                        R.layout.spinner_item, langListPretty);
                        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item);
                        spinner.setAdapter(adapter);
                        spinner.setSelection(uglyList.getPosition(positionListPref));
                    } catch (NullPointerException e) {
                        Log.e(TAG, e.toString());
                    }

                }
            });

        }
    }

}

File: app/src/main/kotlin/org/teamblueridge/pasteitapp/MainActivity.kt
package org.teamblueridge.pasteitapp

import android.Manifest
import android.app.Activity
import android.app.ProgressDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.text.Html
import android.text.method.LinkMovementMethod
import android.util.Log
import android.util.Patterns
import android.view.Menu
import android.view.MenuItem
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import android.widget.ProgressBar
import com.pawegio.kandroid.runAsync
import com.pawegio.kandroid.runOnUiThread
import com.pawegio.kandroid.toast
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.fragment_paste.*
import kotlinx.android.synthetic.main.spinner_item.*
import org.teamblueridge.utils.NetworkUtil
import java.io.*
import java.net.ConnectException
import java.net.URL
import javax.net.ssl.HttpsURLConnection

class MainActivity : AppCompatActivity() {

    private var mReceivedIntent: Intent? = null
    private var mReceivedAction: String? = null
    private val TAG = "TeamBlueRidge"

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        // Set-up the paste fragment and give it a name so we can track it
        if (savedInstanceState == null) {
            fragmentManager.beginTransaction()
                .add(R.id.container, CreatePasteFragment(), "PasteFragment")
                .commit()
        }

        // Set-up up navigation
        fragmentManager.addOnBackStackChangedListener {
            /* Check if we have anything on the stack. If we do, we need to have the back button
             * display in in the ActionBar */
            if (fragmentManager.backStackEntryCount > 0) {
                supportActionBar!!.setHomeButtonEnabled(true)
                supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            } else {
                supportActionBar!!.setHomeButtonEnabled(false)
                supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            }
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window = this.window
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            window.statusBarColor = ContextCompat.getColor(this, R.color.blue_700)
        }
    }

    public override fun onStart() {
        super.onStart()
        mReceivedIntent = intent
        mReceivedAction = mReceivedIntent!!.action
        if (mReceivedAction == Intent.ACTION_VIEW || mReceivedAction == Intent.ACTION_EDIT) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat
                    .requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 0)
            }
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
                == PackageManager.PERMISSION_GRANTED) {
                loadFile()
            } else {
                toast(getString(R.string.request_permissions))
                ActivityCompat.requestPermissions(this,
                    arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 0)
                loadFile()
            }
            mReceivedAction = Intent.ACTION_DEFAULT
        } else {
            paste_content_edittext.setText(mReceivedIntent!!.getStringExtra(Intent.EXTRA_TEXT))
        }

        // Check if the "languages" file exists
        if (File("languages").exists()) {
            populateSpinner()
        } else {
            val prefs = PreferenceManager.getDefaultSharedPreferences(this@MainActivity)
            ApiHandler().getLanguages(this as Activity, UploadDownloadUrlPrep().prepUrl(prefs,
                                      UploadDownloadUrlPrep.DOWNLOAD_LANGS));
            populateSpinner()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_paste -> {
            val view = this.currentFocus
            if (view != null) {
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(view.windowToken, 0)
            }
            prepForPaste()
            true
        }
        R.id.action_settings -> {
            startActivity(Intent(this, SettingsActivity::class.java))
            true
        }
        R.id.action_list -> true
        android.R.id.home -> {
            fragmentManager.popBackStack()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    /**
     * Does some preparation before finally calling the UploadPaste ASyncTask
     */
    fun prepForPaste() {
        val languages = ApiHandler().getLanguages(applicationContext)
        val chosenLang = languages.getKeyByValue(language_spinner.selectedItem) ?: return
        if (!paste_content_edittext.text.toString().isEmpty()) {
            if (NetworkUtil.isConnectedToNetwork(this)) {
                if (chosenLang != "0") {
                    val success = doPaste(if (!chosenLang.isEmpty()) chosenLang else "text")
                    if (success) toast(getString(R.string.paste_toast))
                    paste_name_edittext.setText("")
                    paste_content_edittext.setText("")
                } else {
                    toast(getString(R.string.invalid_language))
                }
            } else {
                toast(getString(R.string.no_network))
            }
        } else {
            toast(getString(R.string.paste_no_text))
        }
    }

    fun <K, V> Map<K, V>.getKeyByValue(value: V): K? {
        for ((k, v) in this) if (v == value) return k
        return null
    }

    /**
     * Uploads the paste to the server then receives the URL of the paste and displays it in the
     * Paste URL Label.
     *
     * @param receivedLanguage  Language to upload the paste in
     */
    fun doPaste(receivedLanguage: String): Boolean {
        val httpUserAgent = ("Paste It v${getString(R.string.version_name)}" +
            ", an Android app for pasting to Stikked (https://goo.gl/LmVtEC)")
        val prefs = PreferenceManager.getDefaultSharedPreferences(this@MainActivity)
        val pDialogUpload = ProgressDialog(this@MainActivity)
        val language = if (!receivedLanguage.isEmpty()) receivedLanguage else "text"
        val title = paste_name_edittext.text.toString()
        val text = paste_content_edittext.text.toString()
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

        //Ensure username is set, if not, default to "mobile user"
        val mUserName =
            if (!prefs.getString("pref_name", "").isEmpty()) prefs.getString("pref_name", "")
            else  "Mobile User " + getString(R.string.version_name)


        pDialogUpload.setMessage(getString(R.string.paste_upload))
        pDialogUpload.isIndeterminate = false
        pDialogUpload.setCancelable(true)
        runOnUiThread {
            pDialogUpload.show()
        }
        var success: Boolean = false
        runAsync {
            // HTTP Header data
            if (!NetworkUtil.isHostAvailable(this, prefs.getString("pref_domain", ""))) {
                pDialogUpload.cancel()
                runOnUiThread {
                    toast("Unable to reach paste server. Check settings and try again.")
                }
                return@runAsync
            }
            val url = URL(UploadDownloadUrlPrep().prepUrl(prefs, UploadDownloadUrlPrep.UPLOAD_PASTE))
            val urlConnection = url.openConnection() as HttpsURLConnection
            urlConnection.requestMethod = "POST"
            urlConnection.doInput = true
            urlConnection.doOutput = true
            urlConnection.setRequestProperty("User-Agent", httpUserAgent)
            val builder = Uri.Builder()
                    .appendQueryParameter("title", title)
                    .appendQueryParameter("text", text)
                    .appendQueryParameter("name", mUserName)
                    .appendQueryParameter("lang", language)
            val query = builder.build().encodedQuery
            urlConnection.outputStream.use {
                val writer = BufferedWriter(OutputStreamWriter(it, "UTF-8"))
                writer.write(query)
                writer.flush()
                writer.close()
                it.close()
            }

            //Get the URL of the paste
            val urlStringBuilder = StringBuilder()
            BufferedReader(InputStreamReader(urlConnection.inputStream))
                    .forEachLine { urlStringBuilder.append(it) }

            val pasteUrl = urlStringBuilder.toString()
            pDialogUpload.dismiss()
            if (Patterns.WEB_URL.matcher(pasteUrl).matches())
                clipboard.primaryClip = ClipData.newPlainText("PasteIt", pasteUrl)

            val labelText: CharSequence
            val toastText: String
            if (Patterns.WEB_URL.matcher(pasteUrl).matches()) {
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
                    labelText = Html.fromHtml("<a href=\"$pasteUrl\">$pasteUrl</a>", Html.FROM_HTML_MODE_LEGACY)
                } else {
                    labelText = Html.fromHtml("<a href=\"$pasteUrl\">$pasteUrl</a>")
                }
                toastText = ""
                success = true
            } else if (pasteUrl == "Invalid API key") {
                toastText = getString(R.string.invalid_api_key)
                labelText = ""
                success = false
            } else {
                Log.e(TAG, "Bad URL: URL received was $pasteUrl")
                labelText = getString(R.string.invalid_url)
                toastText = ""
                success = false
            }

            runOnUiThread {
                paste_url_label.text = labelText
                if (success) {
                    paste_url_label.movementMethod = LinkMovementMethod.getInstance()
                } else {
                    toast(toastText)
                }
            }
        }
        return success
    }

    /**
     * Loads the file being opened and puts it into the paste content EditText.
     */
    fun loadFile() {
        val pDialogFileLoad = ProgressDialog(this@MainActivity)
        pDialogFileLoad.setMessage(getString(R.string.file_load))
        pDialogFileLoad.isIndeterminate = true
        pDialogFileLoad.setCancelable(false)
        pDialogFileLoad.show()
        runAsync {
            mReceivedAction = intent.action
            val receivedText = mReceivedIntent!!.data
            val inputStream = contentResolver.openInputStream(receivedText)
            val reader = BufferedReader(InputStreamReader(inputStream))
            val stringBuilder = StringBuilder()
            reader.forEachLine { stringBuilder.append(it + "\n") }
            inputStream.close()
            runOnUiThread { paste_content_edittext.setText(stringBuilder.toString()) }
        }
        pDialogFileLoad.dismiss()
    }

    /**
     * Populates the spinner for language selection using the arrays of languages generated by
     * the API handler.
     */
    fun populateSpinner() {
        val languages = ApiHandler().getLanguages(applicationContext)
        val prefs = PreferenceManager.getDefaultSharedPreferences(this@MainActivity)
        val positionListPref = prefs.getString("pref_default_language", "text")
        val adapter = ArrayAdapter(applicationContext, R.layout.spinner_item, languages.values.toList())
        adapter.setDropDownViewResource(R.layout.spinner_dropdown_item)
        language_spinner.adapter = adapter
        language_spinner.setSelection(languages.keys.toList().indexOf(positionListPref))
    }

}